software/oada_upload/src/oada_upload.ts

```diff
@@ -0,0 +1,387 @@
+    // Wait 1s for the data to start to refill the db and requery
+    if (!gps.rows.length) {
+      console.log('No unsent data found in database');
+      await sleep(1000);
+      continue;
+    }
+    
+    console.info(`Found batch of ${gps.rows.length} GPS points`);
+    console.debug(`Head time: ${gps.rows[0].time_tz}`);
+
+    // Create DataIndex object(?) for filling data in
+    const data: DataIndex = {};
+    // For each datapoint used, extract data and begin to sort them into buckets to be uploaded
+    // Due to how the OADA server is setup, we the buckets will be by hour collected
+    gps.rows.forEach((p) => {
+      //console.debug(`Processing GPS time: ${p.time_tz}`);
+
+      // Extract time and use it to create path that it will be uploaded it
+      const t = moment.unix(p.time_epoch);
+      const day = t.format('YYYY-MM-DD');
+      const hour = t.format('HH');
+      const path = `/bookmarks/isoblue/device-index/${id}/location/day-index/${day}/hour-index/${hour}`;
+      // Create a UUID that can be coarsely sorted by time of creation
+      // TODO: find a way to use the timestamp to create the id instead of the current time
+      const pId = ksuid.randomSync().string;
+
+      // If the current bucket does not exist, create it
+      if (!data[path]) {
+        data[path] = {
+          epochs: [],
+          timetzs: [],
+          locations: {},
+        };
+      }
+
+      // Insert our data into the bucket
+      data[path].epochs.push(p.time_epoch);
+      data[path].timetzs.push(p.time_tz);
+      data[path].locations[pId] = {
+        id: pId,
+        time: {
+          value: p.time_epoch,
+        },
+        location: {
+          lat: p.lat,
+          lng: p.lng,
+        },
+      };
+      
+    });
+    // pEachSeries executes the following code for each bucket (path) that we created sequentially 
+    await pEachSeries(Object.keys(data), async (path) => {
+      var update_success = true;
+      var res;
+
+      // Attempt the put to OADA
+      try {
+        //console.debug(`GPS ID ${data[path].epochs.join(',')} to OADA ${path}`);
+        console.debug(`${data[path].epochs.length} points to OADA ${path}`);
+        res = await oada.put({
+          tree: isoblueDataTree,
+          path,
+          data: JSON.parse(JSON.stringify({data: data[path].locations})),
+        });
+        //console.debug(`oada put finished: `, res);
+        console.debug(`OADA put finished`);
+      } catch (e) {
+        // The PUT is not a success. This is either due to a misconfiguration or lack of internet.
+        // Assume that everything is configured correctly rebuild the queue for the next try.
+        console.error(`Error Uploading to OADA: %p`, e, e.message, (<Error>e).message, ` `, res);
+        update_success = false;
+      }
+      console.debug(`Done uploading to OADA`);
+      if (update_success){ // Replace the if with a 'continue' in the previous error state?
```

Comment:
This looks right, but I am a bit confused  by it. Why not put a `continue` on line 328 and just drop the conditional here? I am mostly asking to make sure I didn't miss some nuance

Reply:
Wait, I understand the comment now :) ... It's suggesting the same thing as my comment.

Reply:
Done. Needed a `return` instead of a `continue` because it is not a loop